components/collector/src/metric_collectors/change_failure_rate.py

```diff
@@ -0,0 +1,78 @@
+"""Change failure rate collector."""
+
+from collections.abc import Sequence
+from datetime import datetime
+from typing import ClassVar
+
+from base_collectors import MetricCollector
+from collector_utilities.date_time import parse_datetime
+from model import Entities, Entity, MetricMeasurement, SourceMeasurement
+
+
+class ChangeFailureRate(MetricCollector):
+    """Change failure rate collector."""
+
+    # Partial sources for issues and deployments, which need to be combined within this metric
+    ISSUE_SOURCE_TYPES: ClassVar[list[str]] = ["jira"]
+    DEPLOYMENT_SOURCE_TYPES: ClassVar[list[str]] = ["gitlab", "jenkins"]
+
+    async def collect(self) -> MetricMeasurement | None:
+        """Override to calculate the composite value."""
+        if (measurement := await super().collect()) is None:
+            return None
+        deployments = self.deployment_entities(measurement.sources)
+        deployments.sort(key=lambda d: self.deployment_timestamp(d), reverse=True)  # newest to oldest
+        issues = self.issue_entities(measurement.sources)
+        for deployment in deployments:
+            issue_length = len(issues)
+            if issue_length == 0:
+                break  # no more issues; there are no more failed deployments
+            issues = Entities(
+                issue for issue in issues if self.deployment_timestamp(deployment) > self.issue_timestamp(issue)
+            )
+            if len(issues) != issue_length:
+                deployment["failed"] = True  # there were issues after this deployment, so it counts as failed
```

Comment:
I don't understand how this works: you iterate through the deployments from new to old with the goal to find out which deployments failed. I assume a deployment is considered failed if there are issues between the deployment and the next deployment (if any), but I don't see that happening in the code. Moreover, the issue filtering in line 31 seems to find the issues *before* the deployment. I'd expect this loop to look something like this:

```suggestion
        next_timestamp = datetime.max
        for deployment in deployments:
            this_timestamp = self.deployment_timestamp(deployment)
            deployment["failed"] = bool(issues_between(issues, this_timestamp, next_timestamp))
            next_timestamp = this_timestamp
```

And:

```python
    @staticmethod
    def issues_between(issues, timestamp1, timestamp2) -> Entities:
        """Return the issues between the two timestamps."""
        return Entities(issue for issue in issues if timestamp1 < self.issue_timestamp(issue) <= timestamp2)
```

Reply:
The `deployments` are sorted from newest to oldest (line 24), `issues` are only those for which deploy dt > issue dt (line 31). The loop iterates backwards so it can deal with the open ended interval `[deploy_dt -> deploy_dt | inf)`